File: pwiz/pwiz_tools/Topograph/TopographTestProject/BaseTest.cs
/*
 * Original author: Nicholas Shulman <nicksh .at. u.washington.edu>,
 *                  MacCoss Lab, Department of Genome Sciences, UW
 *
 * Copyright 2009 University of Washington - Seattle, WA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.IO;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using pwiz.Topograph.Data;
using pwiz.Topograph.Model;

namespace pwiz.Topograph.Test
{
    public class BaseTest
    {
        private TestContext _testContextInstance;
        /// <summary>
        ///Gets or sets the test context which provides
        ///information about and functionality for the current test run.
        ///</summary>
// ReSharper disable ConvertToAutoProperty
        public TestContext TestContext
        {
            get
            {
                return _testContextInstance;
            }
            set
            {
                _testContextInstance = value;
            }
        }
        // ReSharper restore ConvertToAutoProperty

        public String FindDirectory(String name)
        {
            for (String directory = TestContext.TestDir; null != directory && directory.Length > 10; directory = Path.GetDirectoryName(directory))
            {
                String testDataDirectory = Path.Combine(directory, name);
                if (Directory.Exists(testDataDirectory))
                {
                    return testDataDirectory;
                }
            }
            return null;
        }

        public String GetDataDirectory()
        {
            return FindDirectory("TestData");
        }
        public void CopyDirectory(String sourceDirectory, String destPath)
        {
            // ReSharper disable AssignNullToNotNullAttribute
            String destDirectory = Path.Combine(destPath, Path.GetFileName(sourceDirectory));
            Directory.CreateDirectory(destDirectory);
            foreach (var file in Directory.GetFiles(sourceDirectory))
            {
                File.Copy(file, Path.Combine(destDirectory, Path.GetFileName(file)));
            }
            foreach (var directory in Directory.GetDirectories(sourceDirectory))
            {
                CopyDirectory(directory, destDirectory);
            }
            // ReSharper restore AssignNullToNotNullAttribute
        }

        public void FreshenDirectory(String sourceDirectory, String destDirectory)
        {
            foreach (var file in Directory.GetFiles(sourceDirectory))
            {
                String destFile = Path.Combine(destDirectory, Path.GetFileName(file));
                if (File.Exists(destFile))
                {
                    continue;
                }
                File.Copy(file, destFile);
            }
            foreach (var directory in Directory.GetDirectories(sourceDirectory))
            {
                if (Directory.Exists(Path.Combine(destDirectory, Path.GetFileName(directory))))
                {
                    continue;
                }
                CopyDirectory(directory, destDirectory);
            }
        }

        public void SetupForPwiz()
        {
            String destDirectory = Path.Combine(FindDirectory("TopographTestProject"), "bin\\x86\\Debug");
            FreshenDirectory(Path.Combine(FindDirectory("turnover_lib"), "bin\\x86\\Debug"), destDirectory);
        }

        protected Workspace CreateWorkspace(string path, DbTracerDef dbTracerDef)
        {
            using (var sessionFactory = SessionFactoryFactory.CreateSessionFactory(path, SessionFactoryFlags.CreateSchema))
            {
                using (var session = sessionFactory.OpenSession())
                {
                    var transaction = session.BeginTransaction();
                    var dbWorkspace = new DbWorkspace
                    {
                        ModificationCount = 1,
                        TracerDefCount = dbTracerDef == null ? 0 : 1,
                        SchemaVersion = WorkspaceUpgrader.CurrentVersion,
                    };
                    session.Save(dbWorkspace);
                    if (dbTracerDef != null)
                    {
                        dbTracerDef.Workspace = dbWorkspace;
                        dbTracerDef.Name = "Tracer";
                        session.Save(dbTracerDef);
                    }

                    var modification = new DbModification
                    {
                        DeltaMass = 57.021461,
                        Symbol = "C",
                        Workspace = dbWorkspace
                    };
                    session.Save(modification);
                    transaction.Commit();
                }

            }
            var workspace = new Workspace(path);
            workspace.SetTaskScheduler(TaskScheduler.Default);
            workspace.DatabasePoller.LoadAndMergeChanges(null);
            return workspace;
        }
    }
}

File: pwiz/utility/findmf/base/base/basetest.cpp
//
// $Id$
//
//
// Original author: Witold Wolski <wewolski@gmail.com>
//
// Copyright : ETH Zurich
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#include <boost/cstdint.hpp>

#include "pwiz/utility/findmf/base/base/base.hpp"
#include "pwiz/utility/misc/unit.hpp"

namespace {
  using namespace pwiz::util;
	typedef boost::int32_t int32_t;
  // Tests that the Foo::Bar() method does Abc.
  void testseq() {
    std::vector<double> res;
    ralab::base::base::seq(1.,10.,0.5,res);
    ralab::base::base::seq(10.,2.,res);
    ralab::base::base::seq(2.,10.,res);
    ralab::base::base::seq(10., 2. , -0.5 , res);

    std::vector<int32_t> res2;
    ralab::base::base::seq(10,2,res2);
    ralab::base::base::seq(10,res);
    ralab::base::base::seq(res2,res);

    std::vector<unsigned int> resunsigned;
    ralab::base::base::seq(1u,10u,1u,resunsigned);
    ralab::base::base::seq(1u,10u,resunsigned);

    std::vector<double> resdouble;
    ralab::base::base::seq_length(100. , 1300.,18467,resdouble);
    unit_assert(resdouble.size() == 18467);
    ralab::base::base::seq_length(100. , 1300.,19467,resdouble);
    unit_assert(resdouble.size() == 19467);

    ralab::base::base::seq_length(0.,1000.,1000,resdouble);
    unit_assert(resdouble.size() == 1000);
  }

  // Tests that Foo does Xyz.
  void testmean() {
    std::vector<double> x;
    x.push_back(1.0);
    x.push_back(1.0);
    x.push_back(1.0);
    x.push_back(1.);
    x.push_back(2.);
    x.push_back(3.);
    x.push_back(5.);
    x.push_back(5.);
    x.push_back(6.);
    x.push_back(7.);
    x.push_back(8.);
    double res = ralab::base::base::mean(x);
    unit_assert_equal ( 3.636364, res, 1e-4);
    res = ralab::base::base::mean(x, 0.3);
    unit_assert_equal ( 3.2, res, 1e-4);
    res = ralab::base::base::mean(x, 0.4);
    unit_assert_equal ( 3.33333, res, 1e-4);
    res = ralab::base::base::mean(x, 0.5);
    std::cout << res << std::endl;
    unit_assert_equal ( 3., res, 1e-4);
    res = ralab::base::base::mean(x.begin(),x.end());
    unit_assert_equal ( 3.636364, res, 1e-4);

  }

  void testgeometricmean(){
    std::vector<double> x;
    x.push_back(1.0);
    x.push_back(2.0);
    x.push_back(3.0);

    double res = ralab::base::base::geometricMean(x.begin(), x.end());
    unit_assert_equal( 1.817121, res, 1e-4 );

  }
}  // namespace

int main(int argc, char **argv) {
 testseq();
testmean();
testgeometricmean();
}
